connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/ExecutionHolder.scala

```diff
@@ -0,0 +1,196 @@
+        // In case of exception happening on the handler thread, interrupt the underlying execution.
+        this.interrupt()
+        throw e
+    }
+  }
+
+  protected def execute() = SparkConnectArtifactManager.withArtifactClassLoader {
+    try {
+      // synchronized - check if already got interrupted while starting.
+      synchronized {
+        if (interrupted) {
+          throw new InterruptedException()
+        }
+      }
+
+      session.withActive {
+        val debugString = requestString(executePlanRequest.get)
+
+        session.sparkContext.setJobGroup(
+          jobGroupId,
+          s"Spark Connect - ${StringUtils.abbreviate(debugString, 128)}",
+          interruptOnCancel = true)
+
+        // Add debug information to the query execution so that the jobs are traceable.
+        session.sparkContext.setLocalProperty(
+          "callSite.short",
+          s"Spark Connect - ${StringUtils.abbreviate(debugString, 128)}")
+        session.sparkContext.setLocalProperty(
+          "callSite.long",
+          StringUtils.abbreviate(debugString, 2048))
+
+        executePlanRequest.foreach { request =>
+          request.getPlan.getOpTypeCase match {
+            case proto.Plan.OpTypeCase.COMMAND => handleCommand(request)
+            case proto.Plan.OpTypeCase.ROOT => handlePlan(request)
+            case _ =>
+              throw new UnsupportedOperationException(
+                s"${request.getPlan.getOpTypeCase} not supported.")
+          }
+        }
+      }
+    } catch {
+      // Actually do need to catch Throwable as some failures don't inherit from Exception and
+      // HiveServer will silently swallow them.
+      case e: Throwable =>
+        // scalastyle:off
+        logDebug(s"Exception in execute: $e")
+        // Always cancel all remaining execution after error.
+        sessionHolder.session.sparkContext.cancelJobGroup(jobGroupId)
+        executionError = if (interrupted) {
+          // Turn the interrupt into OPERATION_CANCELLED error.
+          Some(new SparkSQLException("OPERATION_CANCELLED", Map.empty))
+        } else {
+          // Return the originally thrown error.
+          Some(e)
+        }
+    } finally {
+      session.sparkContext.clearJobGroup()
+    }
+  }
+
+  def interrupt(): Unit = {
+    synchronized {
+      interrupted = true
+      if (executionThread != null) {
+        executionThread.interrupt()
+      }
+    }
+  }
+
+  private def handlePlan(request: ExecutePlanRequest): Unit = {
+    val request = executePlanRequest.get
+    val responseObserver = executePlanResponseObserver.get
+
+    val execution = new SparkConnectPlanExecution(this)
+    execution.handlePlan(responseObserver)
+  }
+
+  private def handleCommand(request: ExecutePlanRequest): Unit = {
+    val request = executePlanRequest.get
+    val responseObserver = executePlanResponseObserver.get
+
+    val command = request.getPlan.getCommand
+    val planner = new SparkConnectPlanner(sessionHolder.session)
```

Comment:
I think command execution should be refactored into org.apache.spark.sql.connect.execution.SparkConnectCommandExecution in a followup. It doesn't belong in SparkConnectPlanner.